### PR TITLE
Refactoring move jacoco reports to buildSrc

### PR DESCRIPTION
**Pull request by Avvessalom**
jacoco reports builder moved to own module in buidlSrc

### CODE REVIEW

File: buildSrc/src/main/kotlin/jacocoReport.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+import org.gradle.api.Project
+import org.gradle.testing.jacoco.tasks.JacocoReport
+import java.io.File
+
+fun JacocoReport.jacocoReport(project: Project, moduleName: String, kotlinJsModule: Boolean = false) {
```
Comment (ARTI1208) [inline buildSrc/src/main/kotlin/jacocoReport.kt:29]: `JacocoReport.jacocoReport`.. It feels strange to me, may be rename function to something like `setupReporting`?

File: buildSrc/src/main/kotlin/jacocoReport.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      .walkBottomUp()
+      .toSet()
+
+      classDirectories.setFrom(classFiles)
+      sourceDirectories.setFrom(project.files(coverageSourceDirs))
+      additionalSourceDirs.setFrom(project.files(coverageSourceDirs))
+      executionData
+        .setFrom(project.files("${project.buildDir}/jacoco/jvmTest.exec"))
+
+  }
```
Comment (ARTI1208) [inline buildSrc/src/main/kotlin/jacocoReport.kt:48]:
```suggestion
  }

```

File: projector-agent-common/build.gradle.kts [hunk 3]
```diff
@@ -32,12 +32,7 @@ jacoco {
 }
 
 tasks.withType<JacocoReport> {
-  reports {
-    xml.isEnabled = true
-    xml.destination = file(layout.buildDirectory.dir("../../JacocoReports/jacocoReportAgentCommon.xml"))
-    csv.required.set(false)
-    html.outputLocation.set(layout.buildDirectory.dir("jacocoHtmlProjectorClient"))
-  }
+  jacocoReport(project, "Projector-agent-common")
```
Comment (ARTI1208) [inline projector-agent-common/build.gradle.kts:35]: Why not use previous module camelCase naming (like AgentCommon on line 37 here)? Using kebab-case in module name produces filename with a weird mix of camelCase and kebab-case (jacocoHtmlProjector-agent-common). The same for other module names

**Review by ARTI1208 — CHANGES_REQUESTED** [review]

File: buildSrc/src/main/kotlin/jacocoReport.kt [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+      classDirectories.setFrom(classFiles)
+      sourceDirectories.setFrom(project.files(coverageSourceDirs))
+      additionalSourceDirs.setFrom(project.files(coverageSourceDirs))
+      executionData
+        .setFrom(project.files("${project.buildDir}/jacoco/jvmTest.exec"))
+  }
+  reports {
+    xml.required.set(true)
+    xml.outputLocation.set(project.file(project.layout.buildDirectory.dir("../../JacocoReports/jacocoReport$moduleName.xml")))
```
Comment (SerVB) [inline buildSrc/src/main/kotlin/jacocoReport.kt:50]: I think the following is more concise, but not sure if it works:

```suggestion
    xml.outputLocation.set(project.rootProject.file("JacocoReports/jacocoReport$moduleName.xml"))
```

File: projector-util-loading/build.gradle.kts [hunk 5]
```diff
@@ -1,6 +1,15 @@
 plugins {
   kotlin("jvm")
   `maven-publish`
+  jacoco
+}
+
+jacoco {
+  toolVersion = "0.8.7"
```
Comment (SerVB) [inline projector-util-loading/build.gradle.kts:8]: Can we at least move this version to a constant, or maybe even move this whole `jacoco {}` block to buildSrc?

File: projector-util-loading/build.gradle.kts [hunk 6]
```diff
@@ -1,6 +1,15 @@
...
+}
+
+tasks.withType<JacocoReport> {
+  setupReporting(project, "ProjectorUtilLoading")
```
Comment (SerVB) [inline projector-util-loading/build.gradle.kts:12]: By the way, can we move `tasks.withType<JacocoReport> {` inside `setupReporting`?

**Review by SerVB — CHANGES_REQUESTED** [review]
It can be improved

File: projector-util-loading/build.gradle.kts (hunk 6 above)
Comment (Avvessalom) [reply projector-util-loading/build.gradle.kts:12]: yes, but if we remove this task and leave the only function, all `finalizedBy` tasks won't work

File: projector-util-loading/build.gradle.kts (hunk 5 above)
Comment (Avvessalom) [reply projector-util-loading/build.gradle.kts:8]: > Can we at least move this version to a constant

forgot about it. I will add to gradle.properties

>or maybe even move this whole jacoco {} block to buildSrc?

no, we have to configure jacoco regardless of this task

File: projector-client-common/build.gradle.kts [hunk 7]
```diff
@@ -28,34 +28,16 @@ plugins {
...
-  additionalSourceDirs.setFrom(files(coverageSourceDirs))
-  executionData
-    .setFrom(files("${buildDir}/jacoco/jvmTest.exec"))
-  reports {
-    xml.isEnabled = true
-    xml.destination = file(layout.buildDirectory.dir("../../JacocoReports/jacocoReportClientCommon.xml"))
-    csv.required.set(false)
-    html.outputLocation.set(layout.buildDirectory.dir("jacocoHtmlProjectorClient"))
-  }
+  setupReporting(project, isKotlinMpModule=true)
```
Comment (ARTI1208) [inline projector-client-common/build.gradle.kts:40]:
```suggestion
  setupReporting(project, isKotlinMpModule = true)
```

File: projector-common/build.gradle.kts [hunk 8]
```diff
@@ -29,35 +29,17 @@ plugins {
...
-  additionalSourceDirs.setFrom(files(coverageSourceDirs))
-  executionData
-    .setFrom(files("${buildDir}/jacoco/jvmTest.exec"))
-  reports {
-    xml.isEnabled = true
-    xml.destination = file(layout.buildDirectory.dir("../../JacocoReports/jacocoReportCommon.xml"))
-    csv.required.set(false)
-    html.outputLocation.set(layout.buildDirectory.dir("jacocoHtmlProjectorClient"))
-  }
+  setupReporting(project, isKotlinMpModule=true)
```
Comment (ARTI1208) [inline projector-common/build.gradle.kts:42]:
```suggestion
  setupReporting(project, isKotlinMpModule = true)
```

**Review by ARTI1208 — CHANGES_REQUESTED** [review]

File: projector-common/build.gradle.kts (hunk 8 above)
Comment (SerVB) [reply projector-common/build.gradle.kts]: I would suggest don't bother much with codestyle. We should set up an automatic linter instead.

**Review by SerVB — APPROVED** [review]

**Pull request merged**